presto-main/src/main/java/com/facebook/presto/operator/MergeJoinOperator.java

```diff
@@ -0,0 +1,548 @@
+            }
+        }
+
+        // no output-in-progress match, but there could be incomplete match
+        if (leftMatch.isPresent()) {
+            checkState(rightMatch.isPresent());
+
+            if (leftPage != null) {
+                // look for continuation of match on the left and/or right sides
+
+                if (!findEndOfMatch(leftMatch.get(), leftPage, leftTypes, leftJoinChannels)) {
+                    // continue looking for end of the match
+                    leftPage = null;
+                    return null;
+                }
+
+                if (leftMatch.get().inputs.get(leftMatch.get().inputs.size() - 1) == leftPage) {
+                    leftIndex = leftMatch.get().endIndex;
+                }
+            }
+            else if (finishing) {
+                leftMatch.get().complete = true;
+            }
+            else {
+                // needs more input
+                return null;
+            }
+
+            if (rightPage != null) {
+                if (!findEndOfMatch(rightMatch.get(), rightPage, rightTypes, rightJoinChannels)) {
+                    // continue looking for end of the match
+                    rightPage = null;
+                    return null;
+                }
+
+                if (rightMatch.get().inputs.get(rightMatch.get().inputs.size() - 1) == rightPage) {
+                    rightIndex = rightMatch.get().endIndex;
+                }
+            }
+            else if (noMoreRightInput) {
+                rightMatch.get().complete = true;
+            }
+            else {
+                // needs more input
+                return null;
+            }
+        }
+
+        // there is no output-in-progress match, but there can be a complete match ready for output
+        if (leftMatch.isPresent()) {
+            checkState(leftMatch.get().complete);
+            checkState(rightMatch.isPresent() && rightMatch.get().complete);
+
+            if (addToOutput()) {
+                return buildOutputPage();
+            }
+        }
+
+        if (leftPage == null || rightPage == null) {
+            if (joinType == JoinNode.Type.LEFT) {
+                if (leftPage != null && noMoreRightInput) {
+                    prepareOutput();
+                    while (true) {
+                        if (outputPageBuilder.isFull()) {
+                            return buildOutputPage();
+                        }
+
+                        addOutputRowForLeftJoin();
+
+                        ++leftIndex;
+                        if (leftIndex == leftPage.getPositionCount()) {
+                            // ran out of rows on the left side
+                            leftPage = null;
+                            return null;
+                        }
+                    }
+                }
+
+                if (finishing && outputPageBuilder != null) {
+                    return buildOutputPage();
+                }
+            }
```

Comment:
I saw several places where we have if statement to handle left join, which is not a very good coding structure, we need to make it more concise and straightforward

For now, I would recommend removing scenarios of "left join", let's just support "inner join" for now in the first version and get this PR merged

Let me know if you think differently, else I would also add in a new PR those constraint for query planning

Reply:
I made the changes to make this cleaner. Only supports left join and inner join right now.